src/Microsoft.OData.Core/UriParser/Resolver/UnqualifiedODataUriResolver.cs

```diff
@@ -47,27 +49,55 @@ public override IEnumerable<IEdmOperation> ResolveBoundOperations(IEdmModel mode
                 return base.ResolveBoundOperations(model, identifier, bindingType);
             }
 
-            return FindAcrossModels<IEdmOperation>(model, identifier, this.EnableCaseInsensitive)
-                .Where(operation =>
-                    operation.IsBound
-                    && operation.Parameters.Any()
-                    && operation.HasEquivalentBindingType(bindingType));
+            return FindOperations(model, identifier, this.EnableCaseInsensitive, true, bindingType);
         }
 
-        private static IEnumerable<T> FindAcrossModels<T>(IEdmModel model, String qualifiedName, bool caseInsensitive) where T : IEdmSchemaElement
+        private static IEnumerable<IEdmOperation> FindOperations(IEdmModel model, string qualifiedName, bool caseInsensitive, bool isBound = false, IEdmType bindingType = null)
         {
-            Func<IEdmModel, IEnumerable<T>> finder = (refModel) =>
-                refModel.SchemaElements.OfType<T>()
-                .Where(e => string.Equals(qualifiedName, e.Name, caseInsensitive ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal));
+            IList<IEdmOperation> results;
+            ConcurrentDictionary<string, IList<IEdmOperation>> nameDict;
 
-            IEnumerable<T> results = finder(model);
-
-            foreach (IEdmModel reference in model.ReferencedModels)
+            if (!operationCache.TryGetValue(model, out nameDict))
             {
-                results.Concat(finder(reference));
+                nameDict = new ConcurrentDictionary<string, IList<IEdmOperation>>();
+                operationCache.TryAdd(model, nameDict);
             }
 
+            //Caching safely assuming, case sensitivity for all request will be the same
+            if (!nameDict.TryGetValue(qualifiedName, out results))
+            {
+                results = new List<IEdmOperation>();
+
+                StringComparison strComparison = caseInsensitive ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal;
+
+                GetOperationsForModel(model, qualifiedName, isBound, bindingType, results, strComparison);
+
+                foreach (IEdmModel reference in model.ReferencedModels)
+                {
+                    GetOperationsForModel(reference, qualifiedName, isBound, bindingType, results, strComparison);
+                }
+
+                nameDict.TryAdd(qualifiedName, results);
+                operationCache[model] = nameDict;
+            }
+           
             return results;

```

Comment:
We could cache ReadOnlyCollection or something similar. We should not create a new instance every time though, as that would defeat the purpose of caching.

Reply:
Well, there would still be some performance benefits even if we create a new list each time, but point taken. If you're referring to `System.Collections.ObjectModel.ReadOnlyCollection`, I am not really a fan for the same reason because it has the issue where someone who wants to be able to modify the underlying items can do so through the `Items` property. But, effectively that is the approach I am suggestion: an `IReadOnlyList` implementation that does not offer a way to get at the items that we are storing internally.

Reply:
Items is protected though, but it could be invoked by a caller using reflection for example. I am with you, I meant something that is IReadOnlyList and throws an exception if caller tries to mutate.

Reply:
Thanks for pointing out the `protected` thing, that was a fun 20 minute rabbit hole. I'd forgotten the intricacies of when you can access `private` vs `protected` from within and without the declaring type :)